lang/cs/Org.Apache.REEF.IO/DataCache.cs

```diff
@@ -0,0 +1,276 @@
+        private CacheLevel CacheToDisk()
+        {
+            switch (_cacheLevel)
+            {
+                case CacheLevel.NotLocal:
+                    _diskDirectory = Optional<DirectoryInfo>.Of(_dataMover.RemoteToDisk());
+                    break;
+                default:
+                    throw new SystemException(
+                        "Unexpected cache level transition from " + _cacheLevel + " to " + CacheLevel.Disk);
+            }
+
+            _cacheLevel = CacheLevel.Disk;
+            return _cacheLevel;
+        }
+
+        /// <summary>
+        /// Cache to memory with the injected <see cref="IDataMover{T}"/>.
+        /// </summary>
+        /// <returns>The cached level.</returns>
+        public CacheLevel CacheToMemory()
+        {
+            switch (_cacheLevel)
+            {
+                case CacheLevel.Disk:
+                    CheckDisk();
+                    _memStreams = Optional<IReadOnlyCollection<MemoryStream>>.Of(
+                        new List<MemoryStream>(_dataMover.DiskToMemory(_diskDirectory.Value)));
+                    CleanUpDisk();
+                    break;
+                case CacheLevel.NotLocal:
+                    _memStreams = Optional<IReadOnlyCollection<MemoryStream>>.Of(
+                        new List<MemoryStream>(_dataMover.RemoteToMemory()));
+                    break;
+                default:
+                    throw new SystemException(
+                        "Unexpected cache level transition from " + _cacheLevel + " to " + CacheLevel.InMemoryAsStream);
+            }
+
+            _cacheLevel = CacheLevel.InMemoryAsStream;
+            return _cacheLevel;
+        }
+
+        /// <summary>
+        /// "Materializes" and deserializes the data, with an option to cache.
+        /// </summary>
+        /// <returns>The deserialized data.</returns>
+        public IEnumerable<T> Materialize(bool shouldCache)
+        {
+            if (shouldCache)
+            {
+                CacheToMaterialized();
+            }
+
+            switch (_cacheLevel)
+            {
+                case CacheLevel.InMemoryAsStream:
+                    CheckMemory();
+                    return _dataMover.MemoryToMaterialized(_memStreams.Value);
+                case CacheLevel.Disk:
+                    CheckDisk();
+                    return _dataMover.DiskToMaterialized(_diskDirectory.Value);
+                case CacheLevel.NotLocal:
+                    return _dataMover.RemoteToMaterialized();
+                case CacheLevel.InMemoryMaterialized:
+                    CheckMaterialized();
+                    return _materialized.Value;
+                default:
+                    throw new IllegalStateException("Illegal cache layer.");
+            }
+        }
+
+        /// <summary>
+        /// Deserialize and cache the data in memory from the level the data is currently at.
+        /// </summary>
+        private CacheLevel CacheToMaterialized()
```

Comment:
We are cleaning higher level as we go to lower levels. Is it desirable? I mean consider the case, where user caches data on disk first from remote. He then required some fast operations and as a result cached to memory and did those operations. He then required to do some other memory intensive operations. Hence, decided to delete memory cache of data, but then still wants the disk cache to get access the data. Do we want to give such flexibility to user or not?